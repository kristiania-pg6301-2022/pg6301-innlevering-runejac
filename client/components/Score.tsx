import React from "react";
import { QuestionProps } from "../models/QuestionProps";

export function Score(props: QuestionProps) {
  return (
    <div>
      <p className={"score-txt"} data-testid={"score-status"}>
        {props.correct} / {props.answered} correct answered
      </p>
    </div>
  );
}
