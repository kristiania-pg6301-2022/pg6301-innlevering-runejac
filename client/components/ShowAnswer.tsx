import { Link, Route, Routes } from "react-router-dom";
import React from "react";

export const ShowAnswer = () => {
  return (
    <div className={"show-answer-div"}>
      <Routes>
        <Route
          path={"correct"}
          element={<h1 className={"correct-or-wrong-txt"}>✅ Correct!</h1>}
        />
        <Route
          path={"wrong"}
          element={<h1 className={"correct-or-wrong-txt"}>❌ Wrong</h1>}
        />
      </Routes>
      <ul className={"ul"}>
        <li>
          <Link className={"home"} to={"/question"}>
            More questions
          </Link>
        </li>
        <li>
          <Link className={"home"} to={"/"}>
            Home
          </Link>
        </li>
      </ul>
    </div>
  );
};
