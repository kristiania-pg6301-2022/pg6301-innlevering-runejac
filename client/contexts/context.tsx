import { createContext } from "react";
import { randomQuestion } from "../quiestions-animals";

export const QuestionContext = createContext({ randomQuestion });
