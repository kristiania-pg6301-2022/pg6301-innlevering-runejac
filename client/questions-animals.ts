export interface QuestionAnimals {
  // todo egen info: endret id til ikke-optional, for POST req
  id: number;
  question: string;
  description?: string | null;
  explanation?: string | null;
  tip?: string | null;
  tags?: {
    name: string;
  }[];
  multiple_correct_answers?: "true" | "false";
  difficulty?: string;
  category?: string;
  // answers er enten string eller null
  answers: Record<string, string | null>;
  // correct answers er string, og er enten "true" eller "false"
  correct_answers: Record<string, "true" | "false">;
}

export function randomQuestion(): QuestionAnimals {
  return Questions[Math.trunc(Math.random() * Questions.length)];
}

export function isCorrectAnswer(
  question: QuestionAnimals | undefined,
  answer: string
) {
  return question?.correct_answers[answer + "_correct"] === "true";
}

// Generated from https://quizapi.io/api/v1/questions?category=code&limit=10&tags=JavaScript
// You need to create an API key at https://quizapi.io/ to generate your own questions
export const Questions: QuestionAnimals[] = [
  {
    id: 974,
    question:
      "A carnivorous animal eats flesh, what does a nucivorous animal eat?",
    description: null,
    answers: {
      answer_a: "Nothing",
      answer_b: "Fruit",
      answer_c: "Seaweed",
      answer_d: "Nuts",
    },
    multiple_correct_answers: "false",
    correct_answers: {
      answer_a_correct: "false",
      answer_b_correct: "false",
      answer_c_correct: "false",
      answer_d_correct: "true",
    },
    explanation: null,
    tip: null,
    tags: [{ name: "Animals" }],
    category: "Animal",
    difficulty: "Medium",
  },
  {
    id: 976,
    question: "What color is a polar bear's skin?",
    description: null,
    answers: {
      answer_a: "White",
      answer_b: "Pink",
      answer_c: "Black",
      answer_d: "Green",
    },
    multiple_correct_answers: "false",
    correct_answers: {
      answer_a_correct: "false",
      answer_b_correct: "false",
      answer_c_correct: "true",
      answer_d_correct: "false",
    },
    explanation: null,
    tip: null,
    tags: [{ name: "Animals" }],
    category: "Animal",
    difficulty: "Medium",
  },
  {
    id: 995,
    question: "Cashmere is the wool from which kind of animal?",
    description: null,
    answers: {
      answer_a: "Goat",
      answer_b: "Camel",
      answer_c: "Sheep",
      answer_d: "Llama",
    },
    multiple_correct_answers: "false",
    correct_answers: {
      answer_a_correct: "true",
      answer_b_correct: "false",
      answer_c_correct: "false",
      answer_d_correct: "false",
    },
    explanation: null,
    tip: null,
    tags: [{ name: "Animals" }],
    category: "Animal",
    difficulty: "Medium",
  },
  {
    id: 986,
    question: "What is the fastest animal?",
    description: null,
    answers: {
      answer_a: "Golden Eagle",
      answer_b: "Cheetah",
      answer_c: "Horsefly",
      answer_d: "Peregrine Falcon",
    },
    multiple_correct_answers: "false",
    correct_answers: {
      answer_a_correct: "false",
      answer_b_correct: "false",
      answer_c_correct: "false",
      answer_d_correct: "true",
    },
    explanation: null,
    tip: null,
    tags: [{ name: "Animals" }],
    category: "Animal",
    difficulty: "Medium",
  },
  {
    id: 988,
    question:
      "What is the name for a male bee that comes from an unfertilized egg?",
    description: null,
    answers: {
      answer_a: "Soldier",
      answer_b: "Drone",
      answer_c: "Worker",
      answer_d: "Male",
    },
    multiple_correct_answers: "false",
    correct_answers: {
      answer_a_correct: "false",
      answer_b_correct: "true",
      answer_c_correct: "false",
      answer_d_correct: "false",
    },
    explanation: null,
    tip: null,
    tags: [{ name: "Animals" }],
    category: "Animal",
    difficulty: "Medium",
  },
  {
    id: 989,
    question: "What is the world's longest venomous snake?",
    description: null,
    answers: {
      answer_a: "King Cobra",
      answer_b: "Green Anaconda",
      answer_c: "Inland Taipan",
      answer_d: "Yellow Bellied Sea Snake",
    },
    multiple_correct_answers: "false",
    correct_answers: {
      answer_a_correct: "true",
      answer_b_correct: "false",
      answer_c_correct: "false",
      answer_d_correct: "false",
    },
    explanation: null,
    tip: null,
    tags: [{ name: "Animals" }],
    category: "Animal",
    difficulty: "Medium",
  },
  {
    id: 981,
    question:
      "What dog bread is one of the oldest breeds of dog and has flourished since before 400 BCE.",
    description: null,
    answers: {
      answer_a: "Pugs",
      answer_b: "Bulldogs",
      answer_c: "Boxers",
      answer_d: "Chihuahua",
    },
    multiple_correct_answers: "false",
    correct_answers: {
      answer_a_correct: "true",
      answer_b_correct: "false",
      answer_c_correct: "false",
      answer_d_correct: "false",
    },
    explanation: null,
    tip: null,
    tags: [{ name: "Animals" }],
    category: "Animal",
    difficulty: "Medium",
  },
  {
    id: 984,
    question: "Which of these species is NOT extinct?",
    description: null,
    answers: {
      answer_a: "Japanese sea lion",
      answer_b: "Tasmanian tiger",
      answer_c: "Saudi gazelle",
      answer_d: "Komodo Dragon",
    },
    multiple_correct_answers: "false",
    correct_answers: {
      answer_a_correct: "false",
      answer_b_correct: "false",
      answer_c_correct: "false",
      answer_d_correct: "true",
    },
    explanation: null,
    tip: null,
    tags: [{ name: "Animals" }],
    category: "Animal",
    difficulty: "Medium",
  },
];
