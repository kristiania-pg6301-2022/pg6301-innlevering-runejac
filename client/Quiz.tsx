import React from "react";
import { Route, Routes, useNavigate } from "react-router-dom";
import { useLoader } from "./hooks/useLoader";
import { FrontPage } from "./components/FrontPage";
import { Score } from "./components/Score";
import { ShowAnswer } from "./components/ShowAnswer";
import { getApis } from "./api/getApis";

export interface QuestionProps {
  answered: number | any;
  correct: number | any;
  reloadScore?: any;
  questionApi?: any | { questions: () => Promise<any> } | Promise<any>;
  postAnswerApi?: any;
}

export function MapQuestions(props: QuestionProps) {
  const navigate = useNavigate();
  const {
    loading,
    error,
    data: question,
    reload: reloadQuestions,
  } = useLoader(props.questionApi);

  async function answerHandler(answer: string, id: number) {
    const answerSent: { isCorrect: boolean } = await props.postAnswerApi(
      answer,
      id
    );

    props.reloadScore();
    // for å refreshe "score" så brukes reload fra /api/question/score- kallet
    if (answerSent.isCorrect) {
      navigate("/answer/correct");
    } else {
      navigate("/answer/wrong");
    }
  }

  if (error) {
    return (
      <div className={"error-message"}>
        An error occurred: {error.toString()}
      </div>
    );
  }

  if (!question || loading) return <h2>Loading...</h2>;

  return (
    <>
      <h2>{question.question}</h2>
      <div>
        {Object.keys(question.answers).map((answer: string) =>
          question?.answers[answer] == null ? null : (
            <div key={answer} data-testid={answer}>
              <button
                className={"button"}
                name={"answer"}
                onClick={() => answerHandler(answer, question.id)}
              >
                {question.answers[answer]}
              </button>
            </div>
          )
        )}
      </div>
      <Score
        correct={props.correct}
        answered={props.answered}
        reloadScore={props.reloadScore}
      />
    </>
  );
}

function Quiz() {
  const {
    loading,
    error,
    data: score,
    reload: reloadScore,
  } = useLoader(getApis.scoreApi);

  // måtte ha ? her fordi jeg fikk
  // "TypeError: Cannot read properties of undefined (reading 'correct')"
  const correct = Number(score?.correct);
  const answered = Number(score?.answers);

  return (
    <Routes>
      <Route
        path={"/"}
        element={<FrontPage correct={correct} answered={answered} />}
      />
      <Route
        path={"/question"}
        element={
          <MapQuestions
            correct={correct}
            answered={answered}
            reloadScore={reloadScore}
            questionApi={getApis.questionApi}
            postAnswerApi={getApis.postAnswerApi}
          />
        }
      />
      <Route path={"/answer/*"} element={<ShowAnswer />} />
      <Route path={"*"} element={<h1>Not found</h1>} />
    </Routes>
  );
}

export default Quiz;
