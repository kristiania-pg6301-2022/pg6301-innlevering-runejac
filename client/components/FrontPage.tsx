import { Link } from "react-router-dom";
import React from "react";
import { QuestionProps } from "../Quiz";

export function FrontPage(props: QuestionProps) {
  return (
    <div>
      <h1 className={"correct-or-wrong-txt"}>Take a quiz</h1>
      <h4
        className={"correct-or-wrong-txt"}
        data-testid={"frontpage-score-status"}
      >
        {props.correct} / {props.answered} correct answered{" "}
      </h4>
      <div>
        <Link className={"home"} to={"/question"}>
          New quiz
        </Link>
      </div>
    </div>
  );
}
