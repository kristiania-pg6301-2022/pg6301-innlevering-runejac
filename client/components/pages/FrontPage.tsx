import { Link } from "react-router-dom";
import React from "react";
import { QuestionProps } from "../../Quiz";
import { Score } from "../Score";

export function FrontPage(props: QuestionProps) {
  return (
    <div>
      <h1 className={"question-and-score-txt"}>Take a quiz</h1>
      <Score correct={props.correct} answered={props.answered} />
      <div>
        <Link className={"home"} to={"/question"}>
          New quiz
        </Link>
      </div>
    </div>
  );
}
