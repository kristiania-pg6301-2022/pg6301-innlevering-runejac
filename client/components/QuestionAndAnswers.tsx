import { useNavigate } from "react-router-dom";
import { useLoader } from "../hooks/useLoader";
import { Score } from "./Score";
import { Home } from "./pages/Home";
import React from "react";
import { QuestionProps } from "../Quiz";

export function QuestionAndAnswers(props: QuestionProps) {
  const navigate = useNavigate();
  const {
    loading,
    error,
    data: question,
    reload: reloadQuestions,
  } = useLoader(props.questionApi);

  async function answerHandler(answer: string, id: number) {
    const answerSent: { isCorrect: boolean } = await props.postAnswerApi!(
      answer,
      id
    );

    props.reloadScore!();
    // for å refreshe "score" så brukes reload fra /api/question/score- kallet
    if (answerSent.isCorrect) {
      navigate("/answer/correct");
    } else {
      navigate("/answer/wrong");
    }
  }

  if (error) {
    return (
      <div className={"error-message"}>
        An error occurred: {error.toString()}
      </div>
    );
  }

  if (!question || loading) return <h2>Loading...</h2>;

  return (
    <>
      <div>
        <p className={"question-and-score-txt"}>{question.question}</p>
        {Object.keys(question.answers).map((answer: string) =>
          question?.answers[answer] == null ? null : (
            <div key={answer} data-testid={answer}>
              <button
                className={"button"}
                name={"answer"}
                onClick={() => answerHandler(answer, question.id)}
              >
                {question.answers[answer]}
              </button>
            </div>
          )
        )}
      </div>
      <Score correct={props.correct} answered={props.answered} />
      <Home />
    </>
  );
}
