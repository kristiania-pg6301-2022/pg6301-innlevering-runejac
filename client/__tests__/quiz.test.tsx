import React from "react";
import Quiz, { MapQuestions } from "../Quiz";
import { render, unmountComponentAtNode } from "react-dom";
import { MemoryRouter } from "react-router-dom";
import { act, Simulate } from "react-dom/test-utils";
import "regenerator-runtime/runtime";
import Mock = jest.Mock;
import { QuestionContext } from "../contexts/context";
import fetch from "node-fetch";
import { FrontPage } from "../components/FrontPage";
import { Score } from "../components/Score";
import { randomQuestion } from "../questions-animals";

describe("quiz pages", () => {
  let container: HTMLDivElement;

  /*   const questionApi = {
    listQuestion: async () => [
      {
        id: 42,
        question: "What is the slowest animal in the" + " world?",
        answers: {
          answer_a: "The turtle",
          answer_b: "The sloth",
        },
        correct_answers: {
          answer_a_correct: "false",
          answer_b_correct: "true",
        },
      },
    ],
  }; */

  const questionApi = () => ({
    question: "What is the slowest animal in the world?",
    id: 42,
    answers: {
      answer_a: "The turtle",
      answer_b: "The sloth",
    },
    correct_answers: {
      answer_a_correct: "false",
      answer_b_correct: "true",
    },
  });

  beforeEach(async () => {
    // setup a DOM element as a render target
    container = await document.createElement("div");
  });

  afterEach(async () => {
    // cleanup on exiting
    unmountComponentAtNode(await container);
    jest.fn().mockClear();
  });

  /**-------------------------------------------------------------------------------------**/

  it("should only render startpage", async function () {
    await act(async () => {
      await render(
        <MemoryRouter>
          <Quiz />
        </MemoryRouter>,
        container
      );
    });
    expect(container.innerHTML).toMatchSnapshot();
  });

  // sjekker Score komponent, sender antall rette og totalt spm besvart
  it("at Score: Should test questions answered and correct answered", function () {
    render(
      <MemoryRouter>
        <Score correct={4} answered={5} />
      </MemoryRouter>,
      container
    );

    expect(
      container.querySelector("[data-testid=score-status]")?.textContent
    ).toEqual("4 / 5 correct answered");
    expect(container.innerHTML).toMatchSnapshot();
  });

  it("at FrontPage: Should test questions answered and correct answered", function () {
    render(
      <MemoryRouter>
        <FrontPage correct={4} answered={5} />
      </MemoryRouter>,
      container
    );
    expect(
      container.querySelector("[data-testid=frontpage-score-status]")
        ?.textContent
    ).toEqual("4 / 5 correct answered");
    expect(container.innerHTML).toMatchSnapshot();
  });

  it("should render a question and its answer options", async function () {
    const answered = jest.fn().mockReturnValue(0);
    const correct = jest.fn().mockReturnValue(0);
    // TODO her må du gjøre deg DONE angående score osv

    await act(async () => {
      render(
        <MemoryRouter initialEntries={["/question"]}>
          <MapQuestions
            answered={answered()}
            correct={correct()}
            questionApi={questionApi}
          />
        </MemoryRouter>,
        container
      );
    });
    expect(container.innerHTML).toMatchSnapshot();
    expect(questionApi).toBeDefined();
    expect(answered).toBeCalled();
    expect(correct).toBeCalled();
  });

  it("should register correct answer with simulate click", async function () {
    const reload = jest.fn();
    const postAnswer = jest.fn().mockImplementation(randomQuestion);
    const answered = (x: number) => x + 1;
    const correct = (x: number) => x + 1;
    // TODO her må du gjøre deg DONE angående score osv

    await act(async () => {
      render(
        <MemoryRouter initialEntries={["/question"]}>
          <MapQuestions
            correct={correct}
            answered={answered}
            reloadScore={reload}
            questionApi={questionApi}
            postAnswerApi={postAnswer}
          />
        </MemoryRouter>,
        container
      );
    });

    await act(async () => {
      Simulate.click(container.querySelector("[data-testid=answer_b] button")!);
    });
    expect(reload).toBeCalled();
    expect(postAnswer).toBeCalled();
    expect(container.innerHTML).toMatchSnapshot();
    expect(
      container.querySelector("[data-testid=score-status]")?.textContent
    ).toEqual(`1 / 1 correct answered`);
  });

  it("should register wrong answer with simulate click", async function () {
    const reload = jest.fn();
    const postAnswer = jest.fn();

    await act(async () => {
      render(
        <MemoryRouter initialEntries={["/question"]}>
          <MapQuestions
            correct={0}
            answered={1}
            reloadScore={reload}
            questionApi={questionApi}
            postAnswerApi={postAnswer}
          />
        </MemoryRouter>,
        container
      );
    });

    await act(async () => {
      Simulate.click(container.querySelector("[data-testid=answer_a] button")!);
    });
    expect(reload).toBeCalled();
    // bruker "not" her fordi den ikke skal bli kalt, fordi svaret er feil og setCorrectAnswered skal ikke kalles
    expect(postAnswer).toBeCalled();
    expect(container.innerHTML).toMatchSnapshot();
    expect(
      container.querySelector("[data-testid=score-status]")?.textContent
    ).toEqual("0 / 1 correct answered");
  });

  // henter ut keys(property) fra et array objekt
  xit("should get keys from an object", function () {
    // TODO HUSK xIT her, dette er ikke noe som passerer fra før
    /* const answerNames: string[] = Object.keys(questionApi).filter(
      (a) => questionApi[a]
    ); */

    const keys = Object.keys(questionApi);

    console.log(keys);
    expect(keys).toEqual(["answer_a", "answer_b"]);
  });

  // henter ut values fra et array objekt
  xit("should get values from an array object", function () {
    // TODO HUSK xIT her, dette er ikke noe som passerer fra før
    const answerValues = Object.values(questionApi).filter(
      (a: string | null) => a !== null
    );
    expect(answerValues).toEqual(["The turtle", "The sloth"]);
  });
});
