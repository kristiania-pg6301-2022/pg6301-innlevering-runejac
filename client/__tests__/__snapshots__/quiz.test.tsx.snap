// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`quiz pages At FrontPage: Should test questions answered and correct answered 1`] = `"<div><h1 class=\\"correct-or-wrong-txt\\">Take a quiz</h1><h4 class=\\"correct-or-wrong-txt\\" data-testid=\\"frontpage-score-status\\">4 / 5 correct answered</h4><div><a class=\\"home\\" href=\\"/question\\">New quiz</a></div></div>"`;

exports[`quiz pages At Score: Should test questions answered and correct answered 1`] = `"<div><p data-testid=\\"score-status\\">4 / 5 correct answered</p><p><a class=\\"home\\" href=\\"/\\">Home</a></p></div>"`;

exports[`quiz pages Should only render startpage 1`] = `"<div><h1 class=\\"correct-or-wrong-txt\\">Take a quiz</h1><h4 class=\\"correct-or-wrong-txt\\" data-testid=\\"frontpage-score-status\\">NaN / NaN correct answered</h4><div><a class=\\"home\\" href=\\"/question\\">New quiz</a></div></div>"`;

exports[`quiz pages should register correct answer with simulate click 1`] = `"<h2>What is the slowest animal in the world?</h2><div><div data-testid=\\"answer_a\\"><button class=\\"button\\" name=\\"answer\\">The turtle</button></div><div data-testid=\\"answer_b\\"><button class=\\"button\\" name=\\"answer\\">The sloth</button></div></div><div><p data-testid=\\"score-status\\">1 / 1 correct answered</p><p><a class=\\"home\\" href=\\"/\\">Home</a></p></div>"`;

exports[`quiz pages should register wrong answer with simulate click 1`] = `"<h2>What is the slowest animal in the world?</h2><div><div data-testid=\\"answer_a\\"><button class=\\"button\\" name=\\"answer\\">The turtle</button></div><div data-testid=\\"answer_b\\"><button class=\\"button\\" name=\\"answer\\">The sloth</button></div></div><div><p data-testid=\\"score-status\\">0 / 1 correct answered</p><p><a class=\\"home\\" href=\\"/\\">Home</a></p></div>"`;

exports[`quiz pages should render a question 1`] = `"<h2>What is the slowest animal in the world?</h2><div><div data-testid=\\"answer_a\\"><button class=\\"button\\" name=\\"answer\\">The turtle</button></div><div data-testid=\\"answer_b\\"><button class=\\"button\\" name=\\"answer\\">The sloth</button></div></div><div><p data-testid=\\"score-status\\">0 / 0 correct answered</p><p><a class=\\"home\\" href=\\"/\\">Home</a></p></div>"`;

exports[`quiz pages should render a question and its answer options 1`] = `"<h2>What is the slowest animal in the world?</h2><div><div data-testid=\\"answer_a\\"><button class=\\"button\\" name=\\"answer\\">The turtle</button></div><div data-testid=\\"answer_b\\"><button class=\\"button\\" name=\\"answer\\">The sloth</button></div></div><div><p data-testid=\\"score-status\\">0 / 0 correct answered</p><p><a class=\\"home\\" href=\\"/\\">Home</a></p></div>"`;
